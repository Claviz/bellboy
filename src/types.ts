import { CoreOptions, UriOptions, UrlOptions } from 'request';
import { RequestPromiseOptions } from 'request-promise';
import { ReadStream } from 'fs';
import { Readable } from 'stream';

export interface IDbConnection {
    user?: string;
    password?: string;
    server?: string;
    host?: string;
    database?: string;
    schema?: string;
}

export interface IReporter {
    report(job: IJob): void;
}

export interface IJobConfig {
    reporters?: IReporter[];
}

export interface IJob {
    on(eventName: string, cb: event): void;
    run(): Promise<void>;
    stop(message?: string): void;
    onAny(cb: event): void;
}

export interface IProcessorConfig {
    rowLimit?: number;
}

export interface IDynamicProcessorConfig extends IProcessorConfig {
    generator: () => AsyncIterableIterator<any>;
}

export interface IDirectoryProcessorConfig extends IProcessorConfig {
    path: string;
    filePattern?: string;
    files?: string[];
}

export interface IExcelProcessorConfig extends IDirectoryProcessorConfig {
    hasHeader?: boolean;
    ignoreEmpty?: boolean;
    sheets?: ((number | string)[] | sheetGetter);
}

export interface IProcessor {
    process(processStream: processStream, ...args: any): Promise<void>;
    rowLimit: number;
}

export interface IDestination {
    loadBatch: (data: any[]) => Promise<void>;
    batchSize: number;
    recordGenerator: ((row: any) => AsyncIterableIterator<{}>) | undefined;
    batchTransformer: ((rows: any[]) => Promise<any[]>) | undefined;
    // loadInMode: boolean;
    disableLoad: boolean;
}

export interface IDestinationConfig {
    disableLoad?: boolean;
    batchSize?: number;
    recordGenerator?: (row: any) => AsyncIterableIterator<{}>;
    batchTransformer?: (rows: any[]) => Promise<any[]>;
}

export interface IDatabaseDestinationConfig extends IDestinationConfig {
    table: string;
    connection: IDbConnection;
}

export interface IPostgresDestinationConfig extends IDatabaseDestinationConfig {
    upsertConstraints?: string[];
}

export interface IMssqlDestinationConfig extends IDatabaseDestinationConfig { }

export interface IHttpDestinationConfig extends IDestinationConfig {
    request: (UriOptions & RequestPromiseOptions) | (UrlOptions & RequestPromiseOptions);
}

export interface IStdoutDestinationConfig extends IDestinationConfig {
    asTable?: boolean;
}

export interface IHttpProcessorConfig extends IProcessorConfig {
    connection: CoreOptions & UrlOptions;
    nextRequest?: (header: any) => Promise<(CoreOptions & UrlOptions) | null | undefined>;
}

export interface IJsonHttpProcessorConfig extends IHttpProcessorConfig {
    dataFormat: 'json';
    jsonPath: string;
}

export interface IDelimitedHttpProcessorConfig extends IHttpProcessorConfig {
    dataFormat: 'delimited';
    delimiter: string;
}

export interface IDatabaseProcessorConfig extends IProcessorConfig {
    connection: any;
    query: string;
}

export interface IJsonProcessorConfig extends IDirectoryProcessorConfig {
    jsonPath: string;
}

export interface IDelimitedProcessorConfig extends IDirectoryProcessorConfig {
    delimiter: string;
}

export interface IMssqlProcessorConfig extends IDatabaseProcessorConfig {
}

export interface IPostgresProcessorConfig extends IDatabaseProcessorConfig {
}

export interface ITailProcessorConfig extends IDirectoryProcessorConfig {
    fromBeginning?: boolean;
}

export interface IMqttProcessorConfig extends IProcessorConfig {
    url: string;
    topics: string[];
}

export type DbTypes = 'postgres' | 'mssql';

export type event = (...args: any) => Promise<any>;

export type sheetGetter = (sheetNames: string[]) => Promise<string[]>;

export type emit = (event: string, ...args: any) => Promise<any>;

export type processStream = (readStream: Readable, ...args: any) => Promise<any>;
