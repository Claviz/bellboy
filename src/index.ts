export { Job } from './job';

export { Processor } from './processors/base/processor';
export { DatabaseProcessor } from './processors/base/database-processor';
export { DirectoryProcessor } from './processors/base/directory-processor';

export { ExcelProcessor } from './processors/excel-processor';
export { PostgresProcessor } from './processors/postgres-processor';
export { MssqlProcessor } from './processors/mssql-processor';
export { HttpProcessor } from './processors/http-processor';
export { DynamicProcessor } from './processors/dynamic-processor';
export { JsonProcessor } from './processors/json-processor';
export { MqttProcessor } from './processors/mqtt-processor';
export { TailProcessor } from './processors/tail-processor';
export { DelimitedProcessor } from './processors/delimited-processor';

export { Destination } from './destinations/base/destination';
export { DatabaseDestination } from './destinations/base/database-destination';

export { HttpDestination } from './destinations/http-destination';
export { MssqlDestination } from './destinations/mssql-destination';
export { PostgresDestination } from './destinations/postgres-destination';
export { StdoutDestination } from './destinations/stdout-destination';

export { getDb } from './utils';

export { Reporter } from './reporters/base/reporter'