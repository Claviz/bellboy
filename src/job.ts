import { Readable } from 'stream';

import { emit, event, IDestination, IJob, IJobConfig, IProcessor, IReporter } from './types';

export class Job implements IJob {

    protected events: { [fn: string]: emit[] } = {};
    protected reporters: IReporter[];
    protected anyEvent: emit[] = [];
    protected stopped = false;

    constructor(protected processor: IProcessor, protected destinations: IDestination[], config: IJobConfig = {}) {
        this.reporters = config.reporters || [];
        for (let i = 0; i < this.reporters.length; i++) {
            this.reporters[i].report(this);
        }
    }

    stop(message?: string) { }

    async run() {
        let readStream: Readable;
        let errorMessage: string | undefined;
        this.stop = (message?: string) => {
            this.stopped = true;
            if (readStream) {
                readStream.destroy();
            }
            errorMessage = message;
        };
        await this.emit('startProcessing', this.processor, this.destinations);
        if (!this.stopped) {
            try {
                await this.processor.process(async (_readStream: Readable, ...args: any[]) => {
                    readStream = _readStream;
                    await this.emit('startProcessingStream', ...args);
                    if (!this.stopped) {
                        const results: any[][] = [];
                        for (let j = 0; j < this.destinations.length; j++) {
                            results[j] = [];
                        }
                        let processedRowCount = 0;
                        for await (const row of readStream) {
                            await this.emit('startProcessingRow', row);
                            if (!this.stopped) {
                                for (let i = 0; i < this.destinations.length; i++) {
                                    const recordGeneratorFn = this.destinations[i].recordGenerator;
                                    if (!recordGeneratorFn) {
                                        results[i].push(row);
                                    } else {
                                        try {
                                            for await (const record of recordGeneratorFn(row)) {
                                                results[i].push(record);
                                                await this.emit('rowGenerated', i, record);
                                            }
                                        } catch (err) {
                                            await this.emit('rowGenerationError', i, row, err);
                                        }
                                    }
                                }
                            }
                            await this.emit('endProcessingRow');
                            for (let i = 0; i < this.destinations.length; i++) {
                                while (this.destinations[i].batchSize && results[i].length >= this.destinations[i].batchSize) {
                                    const destination = this.destinations[i];
                                    const toSend = results[i].splice(0, destination.batchSize);
                                    await this.loadBatch(i, destination, toSend);
                                }
                            }
                            processedRowCount++;
                            if (this.processor.rowLimit === processedRowCount) {
                                await this.flushRows(results);
                                this.stop();
                            }
                        }
                        if (!this.stopped) {
                            await this.flushRows(results);
                        }
                    }
                    await this.emit('endProcessingStream', ...args);
                });
            } catch (err) {
                await this.emit('processingError', err);
            }
        }
        await this.emit('endProcessing');
        if (errorMessage) {
            throw new Error(errorMessage);
        }
    }

    on(eventName: 'endProcessing', fn: (() => Promise<any>)): any
    on(eventName: 'loadedBatch', fn: ((destinationIndex: number, data: any[]) => Promise<any>)): any
    on(eventName: 'processingError', fn: ((error: any) => Promise<any>)): any
    on(eventName: 'loadingBatchError', fn: ((destinationIndex: number, data: any[], error: any) => Promise<any>)): any
    on(eventName: 'loadingBatch', fn: ((destinationIndex: number, data: any[]) => Promise<any>)): any
    on(eventName: 'endLoadingBatch', fn: ((destinationIndex: number) => Promise<any>)): any
    on(eventName: 'transformedBatch', fn: ((destinationIndex: number, transformedRows: any) => Promise<any>)): any
    on(eventName: 'transformingBatchError', fn: ((destinationIndex: number, rows: any[], error: any) => Promise<any>)): any
    on(eventName: 'transformingBatch', fn: ((destinationIndex: number, rows: any[]) => Promise<any>)): any
    on(eventName: 'endTransformingBatch', fn: ((destinationIndex: number) => Promise<any>)): any
    on(eventName: 'endProcessingRow', fn: (() => Promise<any>)): any
    on(eventName: 'rowGenerationError', fn: ((destinationIndex: number, row: any, error: any) => Promise<any>)): any
    on(eventName: 'rowGenerated', fn: ((destinationIndex: number, generatedRow: any) => Promise<any>)): any
    on(eventName: 'startProcessingRow', fn: ((row: any) => Promise<any>)): any
    on(eventName: 'startProcessing', fn: ((processor: IProcessor, destinations: IDestination[]) => Promise<any>)): any
    on(eventName: 'endProcessingStream', fn: ((...args: any) => Promise<any>)): any
    on(eventName: 'startProcessingStream', fn: ((...args: any) => Promise<any>)): any
    on(eventName: string, fn: event) {
        const event = this.events[eventName];
        if (!event) {
            this.events[eventName] = [fn];
        } else {
            this.events[eventName].push(fn);
        }
    }

    onAny(fn: event) {
        this.anyEvent.push(fn);
    }

    protected async flushRows(rows: any[][]) {
        for (let i = 0; i < this.destinations.length; i++) {
            if (rows[i].length) {
                await this.loadBatch(i, this.destinations[i], rows[i]);
            }
        }
    }

    protected async loadBatch(destinationIndex: number, destination: IDestination, data: any[]) {
        if (data.length) {
            if (destination.batchTransformer) {
                await this.emit('transformingBatch', destinationIndex, data);
                if (!this.stopped) {
                    try {
                        data = await destination.batchTransformer(data);
                        await this.emit('transformedBatch', destinationIndex, data);
                    } catch (err) {
                        await this.emit('transformingBatchError', destinationIndex, data, err);
                    }
                }
                await this.emit('endTransformingBatch', destinationIndex);
            }
            if (!destination.disableLoad && !this.stopped) {
                await this.emit('loadingBatch', destinationIndex, data);
                if (!this.stopped) {
                    try {
                        await destination.loadBatch(data);
                        await this.emit('loadedBatch', destinationIndex, data);
                    } catch (err) {
                        await this.emit('loadingBatchError', destinationIndex, data, err);
                    }
                }
                await this.emit('endLoadingBatch', destinationIndex);
            }
        }
    }

    protected async emit(eventName: string, ...args: any) {
        try {
            const fn = this.events[eventName];
            if (fn) {
                for (let i = 0; i < fn.length; i++) {
                    await fn[i].apply(this, args);
                }
            }
            for (let i = 0; i < this.anyEvent.length; i++) {
                await this.anyEvent[i].apply(this, [eventName, ...args]);
            }
        } catch (err) {
            console.error(`Warning. Exception was thrown inside ${eventName} event.`, err);
        }
    }
}
