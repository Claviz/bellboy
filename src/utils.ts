import sql from 'mssql';
import rp from 'request-promise';

import { DbTypes, IDbConnection, IPostgresDestination, IMssqlDestination, IHttpDestination } from './types';

const pgp = require('pg-promise')();

export async function sendRequest(data: any[], destination: IHttpDestination) {
    const config = destination.setup;
    const request = {
        method: config.method,
        uri: config.uri,
        json: true,
    };
    if (config.oneRequestPerBatch) {
        const body = await config.transformBody(data);
        await rp({
            ...request,
            body,
        });
    } else {
        for (let i = 0; i < data.length; i++) {
            const body = await config.transformBody(data[i]);
            await rp({
                ...request,
                body,
            });
        }
    }
}

export async function insertToPostgres(data: any[], destination: IPostgresDestination) {
    const columns: string[] = [];
    const upsertConstraints = destination.setup.upsertConstraints;
    for (let i = 0; i < data.length; i++) {
        for (const key of Object.keys(data[i])) {
            if (!columns.includes(key)) {
                columns.push(key);
            }
        }
    }
    const cs = new pgp.helpers.ColumnSet(columns, { table: destination.setup.table });
    const db = await getDb(destination.setup.connection, 'postgres');
    await db.tx(async (t: any) => {
        let query = pgp.helpers.insert(data, cs);
        if (upsertConstraints && upsertConstraints.length) {
            const columnsString = upsertConstraints.map(x => `"${x}"`).join(',');
            query += ` on conflict(${columnsString}) do update set ` +
                cs.assignColumns({ from: 'EXCLUDED', skip: upsertConstraints });
        }
        return t.batch([
            t.none(query),
        ]);
    });
}

export async function insertToMsSql(data: any[], destination: IMssqlDestination) {
    const db = await getDb(destination.setup.connection, 'mssql') as sql.ConnectionPool;
    const query = await db.request().query(`SELECT TOP(0) * FROM ${destination.setup.table}`);
    let table = (query.recordset as any).toTable(destination.setup.table);
    const columns = table.columns.map((x: any) => x.name);
    const transaction = new sql.Transaction(db);
    await transaction.begin();
    for (let x of data) {
        let args = [];
        for (let ii = 0; ii < columns.length; ii++) {
            args.push(x[columns[ii]]);
        }
        table.rows.add(...args);
    }
    await transaction.request().bulk(table);
    await transaction.commit();
}

const cachedDbConnections = new Map<string, { db: any, close: any }>();
export async function getDb(databaseConfig: IDbConnection, dbType: DbTypes) {
    const dbKey = JSON.stringify(databaseConfig);
    if (cachedDbConnections.has(dbKey)) {
        const dbConnection = cachedDbConnections.get(dbKey);
        if (dbConnection) {
            return dbConnection.db;
        }
    }
    if (dbType === 'mssql') {
        const pool = new sql.ConnectionPool({ ...databaseConfig } as sql.config);
        const db = await pool.connect();
        cachedDbConnections.set(dbKey, {
            db,
            close: pool.close.bind(pool),
        });
        return db;
    } else {
        const db = pgp(databaseConfig);
        cachedDbConnections.set(dbKey, {
            db,
            close: pgp.end,
        });
        return db;
    }
};

export async function closeDbConnection(databaseConfig: IDbConnection) {
    const dbKey = JSON.stringify(databaseConfig);
    if (cachedDbConnections.has(dbKey)) {
        const dbConnection = cachedDbConnections.get(dbKey);
        if (dbConnection) {
            await dbConnection.close();
        }
    }
}