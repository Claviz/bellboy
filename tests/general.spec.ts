import { DynamicProcessor, Job } from '../src';
import { CustomDestination, CustomReporter } from './helpers';

beforeAll(async () => {
});

beforeEach(async () => {
});

afterEach(async () => {
});

afterAll(async () => {
});

it('destination should stop loading when rowLimit is specified and reached', async () => {
    const destination = new CustomDestination();
    const processor = new DynamicProcessor({
        generator: async function* () {
            for (let i = 0; i < 10; i++) {
                yield `test${i}`;
            }
        },
        rowLimit: 3,
    });
    const job = new Job(processor, [destination]);
    await job.run();
    expect(destination.getData()).toEqual([
        'test0',
        'test1',
        'test2',
    ]);
});

it('destination should load everything if rowLimit is 0', async () => {
    const destination = new CustomDestination();
    const processor = new DynamicProcessor({
        generator: async function* () {
            for (let i = 0; i < 33; i++) {
                yield `test${i}`;
            }
        },
        rowLimit: 0,
    });
    const job = new Job(processor, [destination]);
    await job.run();
    expect(destination.getData().length).toEqual(33);
});

it(`destination should respsect row limit even if it is less than batchSize`, async () => {
    const data: any[] = [];
    const destination = new CustomDestination({
        batchSize: 10,
    });
    const processor = new DynamicProcessor({
        generator: async function* () {
            for (let i = 0; i < 100; i++) {
                yield `test${i}`;
            }
        },
        rowLimit: 3,
    });
    const job = new Job(processor, [destination]);
    job.on('startProcessingRow', async (row) => {
        data.push(row);
    });
    await job.run();
    expect(data.length).toEqual(3);
});

it(`destination should be able to generate new rows`, async () => {
    const destination = new CustomDestination({
        recordGenerator: async function* (row) {
            for (let i = 0; i < 3; i++) {
                yield `test${i} generated`;
            }
        }
    });
    const processor = new DynamicProcessor({
        generator: async function* () {
            yield `Hello, world!`;
        },
    });
    const job = new Job(processor, [destination]);
    await job.run();
    expect(destination.getData()).toEqual([
        'test0 generated',
        'test1 generated',
        'test2 generated',
    ]);
});

it(`destination shouldn't load if destination load is disabled`, async () => {
    const destination = new CustomDestination({
        disableLoad: true,
    });
    const processor = new DynamicProcessor({
        generator: async function* () {
            for (let i = 0; i < 10; i++) {
                yield `test${i}`;
            }
        }
    });
    const job = new Job(processor, [destination]);
    await job.run();
    expect(destination.getData()).toEqual([]);
});

it(`destination should load if destination load isn't disabled`, async () => {
    const destination = new CustomDestination({
        disableLoad: false,
    });
    const processor = new DynamicProcessor({
        generator: async function* () {
            for (let i = 0; i < 10; i++) {
                yield `test${i}`;
            }
        }
    });
    const job = new Job(processor, [destination]);
    await job.run();
    expect(destination.getData().length).toEqual(10);
});

it(`reporter should report`, async () => {
    const destination = new CustomDestination();
    const processor = new DynamicProcessor({
        generator: async function* () {
            yield `test`;
        }
    });
    const reporter = new CustomReporter();
    const job = new Job(processor, [destination], { reporters: [reporter] });
    await job.run();
    expect(reporter.getEvents()).toMatchSnapshot();
});