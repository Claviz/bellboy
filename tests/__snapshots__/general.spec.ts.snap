// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reporter should report 1`] = `
Array [
  "startProcessing",
  "startProcessingStream",
  "startProcessingRow",
  "endProcessingRow",
  "loadingBatch",
  "loadedBatch",
  "endLoadingBatch",
  "endProcessingStream",
  "endProcessing",
]
`;
